# Machine Learning Model Using Regression for Prediction of Housing Prices

## The first cell contains all the neccesary libraries that will be used in the model and all the preprocessing activities.

In [1]:
import datetime as dt
import time
import sklearn.metrics as metrics
import numpy as np
import pandas as pd
from catboost import Pool
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import  train_test_split, GridSearchCV
from sklearn.metrics import mean_squared_error, r2_score
from catboost import CatBoostRegressor
import math
from IPython.display import display, HTML
display(HTML("<style>.container { width:100% !important; }</style>"))

## Getting the datasets

In [2]:
train = pd.read_csv(r'C:\Users\Admin\Downloads\Data Science projects\Machine Learning Prediction Model using Regression for Housing Prices\train.csv')
test = pd.read_csv(r'C:\Users\Admin\Downloads\Data Science projects\Machine Learning Prediction Model using Regression for Housing Prices\test.csv')
sub = pd.read_csv(r'C:\Users\Admin\Downloads\Data Science projects\Machine Learning Prediction Model using Regression for Housing Prices\sample_submission.csv')

## Viewing the the train.csv dataset

In [3]:
train

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0        1          60       RL         65.0     8450   Pave   NaN      Reg   
1        2          20       RL         80.0     9600   Pave   NaN      Reg   
2        3          60       RL         68.0    11250   Pave   NaN      IR1   
3        4          70       RL         60.0     9550   Pave   NaN      IR1   
4        5          60       RL         84.0    14260   Pave   NaN      IR1   
...    ...         ...      ...          ...      ...    ...   ...      ...   
1455  1456          60       RL         62.0     7917   Pave   NaN      Reg   
1456  1457          20       RL         85.0    13175   Pave   NaN      Reg   
1457  1458          70       RL         66.0     9042   Pave   NaN      Reg   
1458  1459          20       RL         68.0     9717   Pave   NaN      Reg   
1459  1460          20       RL         75.0     9937   Pave   NaN      Reg   

     LandContour Utilities  ... PoolArea PoolQC  Fence MiscFeature MiscVal  \
0            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
2            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
3            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
4            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
...          ...       ...  ...      ...    ...    ...         ...     ...   
1455         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1456         Lvl    AllPub  ...        0    NaN  MnPrv         NaN       0   
1457         Lvl    AllPub  ...        0    NaN  GdPrv        Shed    2500   
1458         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1459         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   

     MoSold YrSold  SaleType  SaleCondition  SalePrice  
0         2   2008        WD         Normal     208500  
1         5   2007        WD         Normal     181500  
2         9   2008        WD         Normal     223500  
3         2   2006        WD        Abnorml     140000  
4        12   2008        WD         Normal     250000  
...     ...    ...       ...            ...        ...  
1455      8   2007        WD         Normal     175000  
1456      2   2010        WD         Normal     210000  
1457      5   2010        WD         Normal     266500  
1458      4   2010        WD         Normal     142125  
1459      6   2008        WD         Normal     147500  

[1460 rows x 81 columns]

## Viewing the the test.csv dataset

In [4]:
test

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0     1461          20       RH           80    11622   Pave   NaN      Reg   
1     1462          20       RL           81    14267   Pave   NaN      IR1   
2     1463          60       RL           74    13830   Pave   NaN      IR1   
3     1464          60       RL           78     9978   Pave   NaN      IR1   
4     1465         120       RL           43     5005   Pave   NaN      IR1   
...    ...         ...      ...          ...      ...    ...   ...      ...   
1454  2915         160       RM           21     1936   Pave   NaN      Reg   
1455  2916         160       RM           21     1894   Pave   NaN      Reg   
1456  2917          20       RL          160    20000   Pave   NaN      Reg   
1457  2918          85       RL           62    10441   Pave   NaN      Reg   
1458  2919          60       RL           74     9627   Pave   NaN      Reg   

     LandContour Utilities  ... ScreenPorch PoolArea PoolQC  Fence  \
0            Lvl    AllPub  ...         120        0    NaN  MnPrv   
1            Lvl    AllPub  ...           0        0    NaN    NaN   
2            Lvl    AllPub  ...           0        0    NaN  MnPrv   
3            Lvl    AllPub  ...           0        0    NaN    NaN   
4            HLS    AllPub  ...         144        0    NaN    NaN   
...          ...       ...  ...         ...      ...    ...    ...   
1454         Lvl    AllPub  ...           0        0    NaN    NaN   
1455         Lvl    AllPub  ...           0        0    NaN    NaN   
1456         Lvl    AllPub  ...           0        0    NaN    NaN   
1457         Lvl    AllPub  ...           0        0    NaN  MnPrv   
1458         Lvl    AllPub  ...           0        0    NaN    NaN   

     MiscFeature MiscVal MoSold  YrSold  SaleType  SaleCondition  
0            NaN       0      6    2010        WD         Normal  
1           Gar2   12500      6    2010        WD         Normal  
2            NaN       0      3    2010        WD         Normal  
3            NaN       0      6    2010        WD         Normal  
4            NaN       0      1    2010        WD         Normal  
...          ...     ...    ...     ...       ...            ...  
1454         NaN       0      6    2006        WD         Normal  
1455         NaN       0      4    2006        WD        Abnorml  
1456         NaN       0      9    2006        WD        Abnorml  
1457        Shed     700      7    2006        WD         Normal  
1458         NaN       0     11    2006        WD         Normal  

[1459 rows x 80 columns]

## This allows for the output to be seen in full for the columns, you can just change the number if you have a different sized dataset.

In [5]:
pd.set_option("display.max_rows", 81)

## Checking null value percentage column wise

In [6]:
train.isnull().mean()*100

Id                0.000000
MSSubClass        0.000000
MSZoning          0.000000
LotFrontage      17.739726
LotArea           0.000000
Street            0.000000
Alley            93.767123
LotShape          0.000000
LandContour       0.000000
Utilities         0.000000
LotConfig         0.000000
LandSlope         0.000000
Neighborhood      0.000000
Condition1        0.000000
Condition2        0.000000
BldgType          0.000000
HouseStyle        0.000000
OverallQual       0.000000
OverallCond       0.000000
YearBuilt         0.000000
YearRemodAdd      0.000000
RoofStyle         0.000000
RoofMatl          0.000000
Exterior1st       0.000000
Exterior2nd       0.000000
MasVnrType        0.547945
MasVnrArea        0.547945
ExterQual         0.000000
ExterCond         0.000000
Foundation        0.000000
BsmtQual          2.534247
BsmtCond          2.534247
BsmtExposure      2.602740
BsmtFinType1      2.534247
BsmtFinSF1        0.000000
BsmtFinType2      2.602740
BsmtFinSF2        0.000000
B

## Finding out the unique values in each column to decide which columns are not useful in training.

In [7]:
train.nunique().sort_values()

CentralAir          2
Utilities           2
Street              2
Alley               2
BsmtHalfBath        3
LandSlope           3
GarageFinish        3
HalfBath            3
PavedDrive          3
PoolQC              3
FullBath            4
MasVnrType          4
BsmtExposure        4
ExterQual           4
MiscFeature         4
BsmtFullBath        4
Fence               4
KitchenQual         4
BsmtCond            4
Fireplaces          4
LandContour         4
LotShape            4
KitchenAbvGr        4
BsmtQual            4
FireplaceQu         5
Electrical          5
YrSold              5
GarageCars          5
GarageQual          5
GarageCond          5
HeatingQC           5
ExterCond           5
MSZoning            5
LotConfig           5
BldgType            5
BsmtFinType2        6
Foundation          6
RoofStyle           6
SaleCondition       6
GarageType          6
BsmtFinType1        6
Heating             6
Functional          7
RoofMatl            8
HouseStyle          8
Condition2

## Find the data types in the dataset in each column to identify the categorical data.

In [8]:
train.dtypes.sort_values()

Id                 int64
BsmtUnfSF          int64
TotalBsmtSF        int64
1stFlrSF           int64
2ndFlrSF           int64
LowQualFinSF       int64
GrLivArea          int64
BsmtFullBath       int64
BsmtHalfBath       int64
FullBath           int64
HalfBath           int64
BedroomAbvGr       int64
KitchenAbvGr       int64
TotRmsAbvGrd       int64
Fireplaces         int64
GarageCars         int64
GarageArea         int64
WoodDeckSF         int64
OpenPorchSF        int64
EnclosedPorch      int64
3SsnPorch          int64
ScreenPorch        int64
PoolArea           int64
MiscVal            int64
MoSold             int64
YrSold             int64
BsmtFinSF2         int64
BsmtFinSF1         int64
SalePrice          int64
MSSubClass         int64
LotArea            int64
OverallQual        int64
OverallCond        int64
YearBuilt          int64
YearRemodAdd       int64
GarageYrBlt      float64
LotFrontage      float64
MasVnrArea       float64
Foundation        object
GarageFinish      object


## Finding the total number of null values in the columns in the dataset.

In [9]:
train.isnull().sum()

Id                  0
MSSubClass          0
MSZoning            0
LotFrontage       259
LotArea             0
Street              0
Alley            1369
LotShape            0
LandContour         0
Utilities           0
LotConfig           0
LandSlope           0
Neighborhood        0
Condition1          0
Condition2          0
BldgType            0
HouseStyle          0
OverallQual         0
OverallCond         0
YearBuilt           0
YearRemodAdd        0
RoofStyle           0
RoofMatl            0
Exterior1st         0
Exterior2nd         0
MasVnrType          8
MasVnrArea          8
ExterQual           0
ExterCond           0
Foundation          0
BsmtQual           37
BsmtCond           37
BsmtExposure       38
BsmtFinType1       37
BsmtFinSF1          0
BsmtFinType2       38
BsmtFinSF2          0
BsmtUnfSF           0
TotalBsmtSF         0
Heating             0
HeatingQC           0
CentralAir          0
Electrical          1
1stFlrSF            0
2ndFlrSF            0
LowQualFin

## Filling in all the null values in the datasets, both the test and the train dataset.

In [10]:
train.fillna(-999, inplace=True)
test.fillna(-999,inplace=True)

## Checking if all the null values in the train dataset are filled.

In [11]:
train.isnull().sum()

Id               0
MSSubClass       0
MSZoning         0
LotFrontage      0
LotArea          0
Street           0
Alley            0
LotShape         0
LandContour      0
Utilities        0
LotConfig        0
LandSlope        0
Neighborhood     0
Condition1       0
Condition2       0
BldgType         0
HouseStyle       0
OverallQual      0
OverallCond      0
YearBuilt        0
YearRemodAdd     0
RoofStyle        0
RoofMatl         0
Exterior1st      0
Exterior2nd      0
MasVnrType       0
MasVnrArea       0
ExterQual        0
ExterCond        0
Foundation       0
BsmtQual         0
BsmtCond         0
BsmtExposure     0
BsmtFinType1     0
BsmtFinSF1       0
BsmtFinType2     0
BsmtFinSF2       0
BsmtUnfSF        0
TotalBsmtSF      0
Heating          0
HeatingQC        0
CentralAir       0
Electrical       0
1stFlrSF         0
2ndFlrSF         0
LowQualFinSF     0
GrLivArea        0
BsmtFullBath     0
BsmtHalfBath     0
FullBath         0
HalfBath         0
BedroomAbvGr     0
KitchenAbvGr

## First,we separate the train.csv dataset by getting a separate collection of the intended output for training the model. Dividing the train dataset to train the model. I also dropped some irrelevant data columns.

In [12]:
X = train.drop(["Id","Alley","PoolQC","Fence","MiscFeature","SalePrice"], axis=1)
Y = train.SalePrice

## Here, we are verifying the datatypes to ensure you get all the categorical data types. No preprocessing activities will be done on the categorical data, CatBoostClassifier will handle all of that to change the categorical data to numerical data to be used by the classification algorithm. 
## Categorical data in simple terms refers to data that refers to specific categories signified by words thus they can be converted to numeric data. words like "yes","no" and "maybe" are converted to 1, 2 and 3 instead. Machine Learning algorithms use mathematical calculations to calculate the predicitions behind the scenes so they need numbers to work, herego the "yes", "no" and "maybe" data is changed to numbers to be able to be used in the mathematical equation.

In [13]:
X.dtypes.sort_values()

MSSubClass         int64
2ndFlrSF           int64
Fireplaces         int64
TotRmsAbvGrd       int64
KitchenAbvGr       int64
BedroomAbvGr       int64
BsmtFinSF1         int64
HalfBath           int64
BsmtFinSF2         int64
BsmtUnfSF          int64
TotalBsmtSF        int64
FullBath           int64
BsmtHalfBath       int64
BsmtFullBath       int64
GrLivArea          int64
1stFlrSF           int64
GarageCars         int64
GarageArea         int64
LowQualFinSF       int64
YearBuilt          int64
YrSold             int64
LotArea            int64
MoSold             int64
MiscVal            int64
PoolArea           int64
ScreenPorch        int64
3SsnPorch          int64
EnclosedPorch      int64
OpenPorchSF        int64
YearRemodAdd       int64
WoodDeckSF         int64
OverallQual        int64
OverallCond        int64
MasVnrArea       float64
GarageYrBlt      float64
LotFrontage      float64
KitchenQual       object
GarageCond        object
Functional        object
GarageQual        object


In [14]:
Y.dtypes


dtype('int64')

## This is the part where you identify all the categorical columns for CatBoost to do its thing.

In [15]:
cat_features = ["Foundation", "GarageFinish", "LandSlope", "LotConfig", "GarageQual", "GarageCond", "LandContour", "Utilities", "GarageType", "LotShape", "Street", "MSZoning", "SaleType", "PavedDrive", "FireplaceQu", "Condition1", "Functional", "BsmtQual", "BsmtCond", "BsmtExposure", "BsmtFinType1", "ExterQual", "BsmtFinType2", "MasVnrType", "Exterior2nd", "Heating", "Neighborhood", "SaleCondition", "Electrical", "Exterior1st", "RoofMatl", "RoofStyle", "HouseStyle", "BldgType", "Condition2", "KitchenQual", "ExterCond", "CentralAir", "HeatingQC"]

## Dividing the train.csv dataset to train the model. You create a training dataset and a testing dataset from the train.csv dataset. The 0.20 signifies the division. So 80% for training and 20% for testing.

In [16]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=.20, random_state=54)

## Here I use the GridSearchCV to check the best parameters to use in the model.

In [25]:
start_time = time.time()
cat_model = CatBoostRegressor()
parameters = {'depth': [2,4,6],'learning_rate' : [0.01,0.05,0.1],'iterations' : [1000,2000,5000]}
grid = GridSearchCV(estimator=cat_model, param_grid = parameters, cv = 2, n_jobs=-1)
grid.fit(X_train, Y_train, cat_features = cat_features, plot = True)
print("--- %s seconds ---" % (time.time() - start_time))

print(" Results from Grid Search " )
print("\n The best estimator across ALL searched params:\n", grid.best_estimator_)
print("\n The best score across ALL searched params:\n", grid.best_score_)
print("\n The best parameters across ALL searched params:\n", grid.best_params_)

MetricVisualizer(layout=Layout(align_self='stretch', height='500px'))

0:	learn: 78273.0604381	total: 20.4ms	remaining: 1m 41s
1:	learn: 75911.6045995	total: 42.5ms	remaining: 1m 46s
2:	learn: 73658.2964609	total: 76.9ms	remaining: 2m 8s
3:	learn: 71551.0364917	total: 105ms	remaining: 2m 10s
4:	learn: 69672.5930756	total: 130ms	remaining: 2m 9s
5:	learn: 67682.7492828	total: 152ms	remaining: 2m 6s
6:	learn: 65676.5117176	total: 200ms	remaining: 2m 22s
7:	learn: 64084.2006646	total: 224ms	remaining: 2m 19s
8:	learn: 62245.5981248	total: 249ms	remaining: 2m 18s
9:	learn: 60557.2218109	total: 279ms	remaining: 2m 19s
10:	learn: 59211.3438117	total: 302ms	remaining: 2m 16s
11:	learn: 57730.0962446	total: 324ms	remaining: 2m 14s
12:	learn: 56405.7573726	total: 349ms	remaining: 2m 14s
13:	learn: 55178.8246945	total: 372ms	remaining: 2m 12s
14:	learn: 53743.0916598	total: 396ms	remaining: 2m 11s
15:	learn: 52564.5041104	total: 420ms	remaining: 2m 10s
16:	learn: 51380.7663485	total: 452ms	remaining: 2m 12s
17:	learn: 50226.1900783	total: 476ms	remaining: 2m 11s
18

149:	learn: 21086.1509875	total: 3.89s	remaining: 2m 5s
150:	learn: 21016.7094277	total: 3.91s	remaining: 2m 5s
151:	learn: 21001.7465239	total: 3.93s	remaining: 2m 5s
152:	learn: 20984.2585901	total: 3.96s	remaining: 2m 5s
153:	learn: 20926.7936283	total: 3.98s	remaining: 2m 5s
154:	learn: 20884.1256871	total: 4.01s	remaining: 2m 5s
155:	learn: 20814.3455514	total: 4.03s	remaining: 2m 5s
156:	learn: 20780.2189907	total: 4.05s	remaining: 2m 5s
157:	learn: 20723.4009969	total: 4.08s	remaining: 2m 5s
158:	learn: 20712.9863941	total: 4.11s	remaining: 2m 4s
159:	learn: 20698.4938305	total: 4.13s	remaining: 2m 4s
160:	learn: 20663.6959568	total: 4.15s	remaining: 2m 4s
161:	learn: 20648.1201389	total: 4.18s	remaining: 2m 4s
162:	learn: 20628.6232828	total: 4.2s	remaining: 2m 4s
163:	learn: 20587.6983289	total: 4.22s	remaining: 2m 4s
164:	learn: 20581.9757653	total: 4.24s	remaining: 2m 4s
165:	learn: 20565.1052771	total: 4.26s	remaining: 2m 4s
166:	learn: 20549.4927628	total: 4.29s	remaining:

304:	learn: 17686.8458950	total: 7.68s	remaining: 1m 58s
305:	learn: 17675.4780069	total: 7.7s	remaining: 1m 58s
306:	learn: 17655.4331004	total: 7.73s	remaining: 1m 58s
307:	learn: 17649.4257632	total: 7.76s	remaining: 1m 58s
308:	learn: 17644.9830844	total: 7.78s	remaining: 1m 58s
309:	learn: 17631.8865766	total: 7.8s	remaining: 1m 58s
310:	learn: 17615.5884764	total: 7.83s	remaining: 1m 58s
311:	learn: 17610.3370172	total: 7.85s	remaining: 1m 57s
312:	learn: 17602.2754929	total: 7.88s	remaining: 1m 57s
313:	learn: 17600.4283879	total: 7.9s	remaining: 1m 57s
314:	learn: 17578.8297214	total: 7.92s	remaining: 1m 57s
315:	learn: 17547.4801781	total: 7.94s	remaining: 1m 57s
316:	learn: 17524.2265592	total: 7.96s	remaining: 1m 57s
317:	learn: 17502.7519201	total: 8s	remaining: 1m 57s
318:	learn: 17497.0304972	total: 8.02s	remaining: 1m 57s
319:	learn: 17490.7675261	total: 8.04s	remaining: 1m 57s
320:	learn: 17464.8239519	total: 8.07s	remaining: 1m 57s
321:	learn: 17462.6557067	total: 8.09

449:	learn: 15436.7516990	total: 11.3s	remaining: 1m 54s
450:	learn: 15406.3255734	total: 11.3s	remaining: 1m 54s
451:	learn: 15391.5879451	total: 11.4s	remaining: 1m 54s
452:	learn: 15366.9090433	total: 11.4s	remaining: 1m 54s
453:	learn: 15344.7129234	total: 11.4s	remaining: 1m 54s
454:	learn: 15335.8527346	total: 11.4s	remaining: 1m 54s
455:	learn: 15327.4049789	total: 11.5s	remaining: 1m 54s
456:	learn: 15315.5341250	total: 11.5s	remaining: 1m 54s
457:	learn: 15310.7383578	total: 11.5s	remaining: 1m 54s
458:	learn: 15296.5749292	total: 11.5s	remaining: 1m 54s
459:	learn: 15286.2347075	total: 11.6s	remaining: 1m 54s
460:	learn: 15285.5003981	total: 11.6s	remaining: 1m 54s
461:	learn: 15282.2825862	total: 11.6s	remaining: 1m 54s
462:	learn: 15268.0471408	total: 11.6s	remaining: 1m 54s
463:	learn: 15257.1067433	total: 11.7s	remaining: 1m 54s
464:	learn: 15254.4688406	total: 11.7s	remaining: 1m 54s
465:	learn: 15208.5778689	total: 11.7s	remaining: 1m 54s
466:	learn: 15190.2252958	total

595:	learn: 14055.8261758	total: 15.2s	remaining: 1m 52s
596:	learn: 14038.5322940	total: 15.3s	remaining: 1m 52s
597:	learn: 14036.0462830	total: 15.3s	remaining: 1m 52s
598:	learn: 14033.6777531	total: 15.3s	remaining: 1m 52s
599:	learn: 14017.5865388	total: 15.3s	remaining: 1m 52s
600:	learn: 14015.7462728	total: 15.3s	remaining: 1m 52s
601:	learn: 14013.2229553	total: 15.4s	remaining: 1m 52s
602:	learn: 14003.9362482	total: 15.4s	remaining: 1m 52s
603:	learn: 13996.5633322	total: 15.4s	remaining: 1m 52s
604:	learn: 13986.4780727	total: 15.5s	remaining: 1m 52s
605:	learn: 13975.2797461	total: 15.5s	remaining: 1m 52s
606:	learn: 13967.6364660	total: 15.5s	remaining: 1m 52s
607:	learn: 13960.5681208	total: 15.5s	remaining: 1m 52s
608:	learn: 13956.9923028	total: 15.5s	remaining: 1m 52s
609:	learn: 13944.9674129	total: 15.6s	remaining: 1m 52s
610:	learn: 13944.1283622	total: 15.6s	remaining: 1m 52s
611:	learn: 13909.7534345	total: 15.6s	remaining: 1m 51s
612:	learn: 13900.6293495	total

747:	learn: 12889.8961607	total: 19.1s	remaining: 1m 48s
748:	learn: 12883.5278343	total: 19.1s	remaining: 1m 48s
749:	learn: 12876.5562497	total: 19.2s	remaining: 1m 48s
750:	learn: 12863.2501489	total: 19.2s	remaining: 1m 48s
751:	learn: 12861.7911740	total: 19.2s	remaining: 1m 48s
752:	learn: 12856.6247091	total: 19.2s	remaining: 1m 48s
753:	learn: 12856.2454062	total: 19.3s	remaining: 1m 48s
754:	learn: 12847.1359415	total: 19.3s	remaining: 1m 48s
755:	learn: 12833.6334654	total: 19.3s	remaining: 1m 48s
756:	learn: 12829.1806446	total: 19.4s	remaining: 1m 48s
757:	learn: 12822.4268135	total: 19.4s	remaining: 1m 48s
758:	learn: 12816.2137932	total: 19.4s	remaining: 1m 48s
759:	learn: 12806.1890685	total: 19.4s	remaining: 1m 48s
760:	learn: 12805.7726162	total: 19.5s	remaining: 1m 48s
761:	learn: 12796.1586977	total: 19.5s	remaining: 1m 48s
762:	learn: 12795.5282241	total: 19.5s	remaining: 1m 48s
763:	learn: 12791.6969842	total: 19.5s	remaining: 1m 48s
764:	learn: 12789.6285431	total

898:	learn: 11922.7404544	total: 23s	remaining: 1m 44s
899:	learn: 11912.2362942	total: 23s	remaining: 1m 44s
900:	learn: 11904.3075484	total: 23s	remaining: 1m 44s
901:	learn: 11899.0952318	total: 23s	remaining: 1m 44s
902:	learn: 11890.1152157	total: 23.1s	remaining: 1m 44s
903:	learn: 11889.5653693	total: 23.1s	remaining: 1m 44s
904:	learn: 11887.7945207	total: 23.1s	remaining: 1m 44s
905:	learn: 11885.3473536	total: 23.2s	remaining: 1m 44s
906:	learn: 11875.8318358	total: 23.2s	remaining: 1m 44s
907:	learn: 11856.1511946	total: 23.2s	remaining: 1m 44s
908:	learn: 11854.9927793	total: 23.2s	remaining: 1m 44s
909:	learn: 11845.8166439	total: 23.3s	remaining: 1m 44s
910:	learn: 11837.0781637	total: 23.3s	remaining: 1m 44s
911:	learn: 11835.5488788	total: 23.3s	remaining: 1m 44s
912:	learn: 11834.7336133	total: 23.3s	remaining: 1m 44s
913:	learn: 11826.5089984	total: 23.4s	remaining: 1m 44s
914:	learn: 11822.2226501	total: 23.4s	remaining: 1m 44s
915:	learn: 11813.0975353	total: 23.4s	

1043:	learn: 11057.8611212	total: 26.8s	remaining: 1m 41s
1044:	learn: 11049.2298483	total: 26.8s	remaining: 1m 41s
1045:	learn: 11031.6284167	total: 26.8s	remaining: 1m 41s
1046:	learn: 11030.7026951	total: 26.8s	remaining: 1m 41s
1047:	learn: 11028.7414133	total: 26.9s	remaining: 1m 41s
1048:	learn: 11022.1835883	total: 26.9s	remaining: 1m 41s
1049:	learn: 11006.7536864	total: 26.9s	remaining: 1m 41s
1050:	learn: 11001.7526669	total: 26.9s	remaining: 1m 41s
1051:	learn: 10998.0380595	total: 27s	remaining: 1m 41s
1052:	learn: 10985.0376507	total: 27s	remaining: 1m 41s
1053:	learn: 10981.5813215	total: 27s	remaining: 1m 41s
1054:	learn: 10976.3963726	total: 27s	remaining: 1m 41s
1055:	learn: 10974.0450547	total: 27.1s	remaining: 1m 41s
1056:	learn: 10969.5116869	total: 27.1s	remaining: 1m 41s
1057:	learn: 10968.6506472	total: 27.1s	remaining: 1m 40s
1058:	learn: 10964.4261179	total: 27.1s	remaining: 1m 40s
1059:	learn: 10960.9350005	total: 27.1s	remaining: 1m 40s
1060:	learn: 10956.972

1192:	learn: 10355.2346423	total: 30.9s	remaining: 1m 38s
1193:	learn: 10353.7868366	total: 30.9s	remaining: 1m 38s
1194:	learn: 10340.7421834	total: 31s	remaining: 1m 38s
1195:	learn: 10335.1722473	total: 31s	remaining: 1m 38s
1196:	learn: 10330.2409620	total: 31s	remaining: 1m 38s
1197:	learn: 10330.2344960	total: 31.1s	remaining: 1m 38s
1198:	learn: 10328.5294583	total: 31.1s	remaining: 1m 38s
1199:	learn: 10328.5233560	total: 31.1s	remaining: 1m 38s
1200:	learn: 10328.1191638	total: 31.1s	remaining: 1m 38s
1201:	learn: 10321.8785251	total: 31.2s	remaining: 1m 38s
1202:	learn: 10321.8735981	total: 31.2s	remaining: 1m 38s
1203:	learn: 10311.5751791	total: 31.2s	remaining: 1m 38s
1204:	learn: 10310.2880459	total: 31.2s	remaining: 1m 38s
1205:	learn: 10309.4185964	total: 31.3s	remaining: 1m 38s
1206:	learn: 10305.7668822	total: 31.3s	remaining: 1m 38s
1207:	learn: 10303.1375585	total: 31.3s	remaining: 1m 38s
1208:	learn: 10302.8637703	total: 31.3s	remaining: 1m 38s
1209:	learn: 10301.0

1337:	learn: 9802.3950742	total: 34.5s	remaining: 1m 34s
1338:	learn: 9800.8351593	total: 34.6s	remaining: 1m 34s
1339:	learn: 9793.3957121	total: 34.6s	remaining: 1m 34s
1340:	learn: 9789.7649972	total: 34.6s	remaining: 1m 34s
1341:	learn: 9789.7532760	total: 34.6s	remaining: 1m 34s
1342:	learn: 9787.6784107	total: 34.7s	remaining: 1m 34s
1343:	learn: 9785.8517472	total: 34.7s	remaining: 1m 34s
1344:	learn: 9785.2735688	total: 34.7s	remaining: 1m 34s
1345:	learn: 9781.9563997	total: 34.7s	remaining: 1m 34s
1346:	learn: 9778.2752002	total: 34.8s	remaining: 1m 34s
1347:	learn: 9771.7349033	total: 34.8s	remaining: 1m 34s
1348:	learn: 9770.6874838	total: 34.8s	remaining: 1m 34s
1349:	learn: 9770.6871232	total: 34.8s	remaining: 1m 34s
1350:	learn: 9770.0419099	total: 34.9s	remaining: 1m 34s
1351:	learn: 9766.1760352	total: 34.9s	remaining: 1m 34s
1352:	learn: 9762.3207267	total: 34.9s	remaining: 1m 34s
1353:	learn: 9757.4140517	total: 35s	remaining: 1m 34s
1354:	learn: 9752.3153488	total: 

1485:	learn: 9298.0969014	total: 38.5s	remaining: 1m 31s
1486:	learn: 9295.3514912	total: 38.5s	remaining: 1m 31s
1487:	learn: 9294.3455073	total: 38.6s	remaining: 1m 31s
1488:	learn: 9293.4805195	total: 38.6s	remaining: 1m 31s
1489:	learn: 9288.2167561	total: 38.6s	remaining: 1m 30s
1490:	learn: 9285.0968483	total: 38.6s	remaining: 1m 30s
1491:	learn: 9275.7879906	total: 38.7s	remaining: 1m 30s
1492:	learn: 9274.7069542	total: 38.7s	remaining: 1m 30s
1493:	learn: 9271.7951292	total: 38.7s	remaining: 1m 30s
1494:	learn: 9265.1109091	total: 38.7s	remaining: 1m 30s
1495:	learn: 9254.5346955	total: 38.8s	remaining: 1m 30s
1496:	learn: 9252.2971735	total: 38.8s	remaining: 1m 30s
1497:	learn: 9250.0658239	total: 38.8s	remaining: 1m 30s
1498:	learn: 9243.4638286	total: 38.8s	remaining: 1m 30s
1499:	learn: 9241.8724773	total: 38.9s	remaining: 1m 30s
1500:	learn: 9241.1198661	total: 38.9s	remaining: 1m 30s
1501:	learn: 9240.1373241	total: 38.9s	remaining: 1m 30s
1502:	learn: 9237.7374328	total

1637:	learn: 8815.6797031	total: 42.6s	remaining: 1m 27s
1638:	learn: 8815.5831547	total: 42.6s	remaining: 1m 27s
1639:	learn: 8812.7190025	total: 42.7s	remaining: 1m 27s
1640:	learn: 8810.7953360	total: 42.7s	remaining: 1m 27s
1641:	learn: 8810.2360340	total: 42.7s	remaining: 1m 27s
1642:	learn: 8808.2940087	total: 42.7s	remaining: 1m 27s
1643:	learn: 8802.7847073	total: 42.8s	remaining: 1m 27s
1644:	learn: 8801.9089142	total: 42.8s	remaining: 1m 27s
1645:	learn: 8797.5031309	total: 42.8s	remaining: 1m 27s
1646:	learn: 8794.1897982	total: 42.9s	remaining: 1m 27s
1647:	learn: 8790.9930674	total: 42.9s	remaining: 1m 27s
1648:	learn: 8786.3228241	total: 42.9s	remaining: 1m 27s
1649:	learn: 8781.0990344	total: 42.9s	remaining: 1m 27s
1650:	learn: 8780.3807790	total: 42.9s	remaining: 1m 27s
1651:	learn: 8774.2623865	total: 43s	remaining: 1m 27s
1652:	learn: 8766.4430634	total: 43s	remaining: 1m 27s
1653:	learn: 8760.1801548	total: 43s	remaining: 1m 27s
1654:	learn: 8760.1800114	total: 43s	

1785:	learn: 8400.1160757	total: 46.8s	remaining: 1m 24s
1786:	learn: 8395.8107646	total: 46.8s	remaining: 1m 24s
1787:	learn: 8395.0814735	total: 46.8s	remaining: 1m 24s
1788:	learn: 8386.0215823	total: 46.8s	remaining: 1m 24s
1789:	learn: 8382.9464009	total: 46.8s	remaining: 1m 24s
1790:	learn: 8379.3713463	total: 46.9s	remaining: 1m 23s
1791:	learn: 8374.1190489	total: 46.9s	remaining: 1m 23s
1792:	learn: 8372.6035011	total: 46.9s	remaining: 1m 23s
1793:	learn: 8370.7524835	total: 47s	remaining: 1m 23s
1794:	learn: 8366.8916253	total: 47s	remaining: 1m 23s
1795:	learn: 8361.3110958	total: 47s	remaining: 1m 23s
1796:	learn: 8355.2313416	total: 47s	remaining: 1m 23s
1797:	learn: 8354.6959380	total: 47s	remaining: 1m 23s
1798:	learn: 8350.8661534	total: 47.1s	remaining: 1m 23s
1799:	learn: 8346.3958335	total: 47.1s	remaining: 1m 23s
1800:	learn: 8344.8537876	total: 47.1s	remaining: 1m 23s
1801:	learn: 8337.6959454	total: 47.1s	remaining: 1m 23s
1802:	learn: 8334.5664359	total: 47.2s	re

1931:	learn: 7960.7072231	total: 50.6s	remaining: 1m 20s
1932:	learn: 7952.6258008	total: 50.6s	remaining: 1m 20s
1933:	learn: 7950.8493772	total: 50.6s	remaining: 1m 20s
1934:	learn: 7942.7168096	total: 50.7s	remaining: 1m 20s
1935:	learn: 7937.7208895	total: 50.7s	remaining: 1m 20s
1936:	learn: 7933.7394864	total: 50.7s	remaining: 1m 20s
1937:	learn: 7931.4388819	total: 50.8s	remaining: 1m 20s
1938:	learn: 7924.4609450	total: 50.8s	remaining: 1m 20s
1939:	learn: 7922.3875083	total: 50.8s	remaining: 1m 20s
1940:	learn: 7920.2256758	total: 50.9s	remaining: 1m 20s
1941:	learn: 7914.9212653	total: 50.9s	remaining: 1m 20s
1942:	learn: 7906.1060271	total: 50.9s	remaining: 1m 20s
1943:	learn: 7904.7876430	total: 50.9s	remaining: 1m 20s
1944:	learn: 7901.7014614	total: 51s	remaining: 1m 20s
1945:	learn: 7897.9337154	total: 51s	remaining: 1m 20s
1946:	learn: 7894.8775898	total: 51s	remaining: 1m 19s
1947:	learn: 7891.9416417	total: 51s	remaining: 1m 19s
1948:	learn: 7890.7254646	total: 51.1s	

2080:	learn: 7546.0832298	total: 54.5s	remaining: 1m 16s
2081:	learn: 7541.4761711	total: 54.6s	remaining: 1m 16s
2082:	learn: 7539.9655322	total: 54.6s	remaining: 1m 16s
2083:	learn: 7538.4509994	total: 54.6s	remaining: 1m 16s
2084:	learn: 7533.7896470	total: 54.6s	remaining: 1m 16s
2085:	learn: 7532.2969205	total: 54.6s	remaining: 1m 16s
2086:	learn: 7531.9319271	total: 54.7s	remaining: 1m 16s
2087:	learn: 7531.2583647	total: 54.7s	remaining: 1m 16s
2088:	learn: 7528.9635613	total: 54.7s	remaining: 1m 16s
2089:	learn: 7527.1615477	total: 54.7s	remaining: 1m 16s
2090:	learn: 7524.7860198	total: 54.8s	remaining: 1m 16s
2091:	learn: 7518.8372171	total: 54.8s	remaining: 1m 16s
2092:	learn: 7516.4075540	total: 54.8s	remaining: 1m 16s
2093:	learn: 7514.9644729	total: 54.8s	remaining: 1m 16s
2094:	learn: 7513.2416222	total: 54.9s	remaining: 1m 16s
2095:	learn: 7509.6319017	total: 54.9s	remaining: 1m 16s
2096:	learn: 7507.3854123	total: 54.9s	remaining: 1m 16s
2097:	learn: 7505.1771290	total

2231:	learn: 7181.4329747	total: 58.5s	remaining: 1m 12s
2232:	learn: 7180.9388601	total: 58.5s	remaining: 1m 12s
2233:	learn: 7180.5752268	total: 58.5s	remaining: 1m 12s
2234:	learn: 7179.5506278	total: 58.6s	remaining: 1m 12s
2235:	learn: 7175.6049492	total: 58.6s	remaining: 1m 12s
2236:	learn: 7172.7816924	total: 58.6s	remaining: 1m 12s
2237:	learn: 7172.5229459	total: 58.7s	remaining: 1m 12s
2238:	learn: 7171.5251360	total: 58.7s	remaining: 1m 12s
2239:	learn: 7163.4657004	total: 58.7s	remaining: 1m 12s
2240:	learn: 7158.1790873	total: 58.7s	remaining: 1m 12s
2241:	learn: 7157.4216439	total: 58.8s	remaining: 1m 12s
2242:	learn: 7154.2367080	total: 58.8s	remaining: 1m 12s
2243:	learn: 7150.9736679	total: 58.8s	remaining: 1m 12s
2244:	learn: 7150.6426882	total: 58.8s	remaining: 1m 12s
2245:	learn: 7148.1070822	total: 58.9s	remaining: 1m 12s
2246:	learn: 7146.5700446	total: 58.9s	remaining: 1m 12s
2247:	learn: 7146.4505056	total: 58.9s	remaining: 1m 12s
2248:	learn: 7139.9658312	total

2382:	learn: 6835.4581672	total: 1m 2s	remaining: 1m 9s
2383:	learn: 6831.9897471	total: 1m 2s	remaining: 1m 9s
2384:	learn: 6830.2824923	total: 1m 2s	remaining: 1m 9s
2385:	learn: 6824.3566936	total: 1m 2s	remaining: 1m 8s
2386:	learn: 6821.0048740	total: 1m 2s	remaining: 1m 8s
2387:	learn: 6819.9240314	total: 1m 3s	remaining: 1m 8s
2388:	learn: 6816.6018783	total: 1m 3s	remaining: 1m 8s
2389:	learn: 6815.9527055	total: 1m 3s	remaining: 1m 8s
2390:	learn: 6810.7480543	total: 1m 3s	remaining: 1m 8s
2391:	learn: 6809.8562346	total: 1m 3s	remaining: 1m 8s
2392:	learn: 6807.4090711	total: 1m 3s	remaining: 1m 8s
2393:	learn: 6805.5428154	total: 1m 3s	remaining: 1m 8s
2394:	learn: 6805.1766106	total: 1m 3s	remaining: 1m 8s
2395:	learn: 6804.1316777	total: 1m 3s	remaining: 1m 8s
2396:	learn: 6803.2260982	total: 1m 3s	remaining: 1m 8s
2397:	learn: 6798.8092700	total: 1m 3s	remaining: 1m 8s
2398:	learn: 6796.8175251	total: 1m 3s	remaining: 1m 8s
2399:	learn: 6793.6991365	total: 1m 3s	remaining

2532:	learn: 6538.1803081	total: 1m 7s	remaining: 1m 5s
2533:	learn: 6538.0660614	total: 1m 7s	remaining: 1m 5s
2534:	learn: 6535.3848040	total: 1m 7s	remaining: 1m 5s
2535:	learn: 6532.9863457	total: 1m 7s	remaining: 1m 5s
2536:	learn: 6531.5486572	total: 1m 7s	remaining: 1m 5s
2537:	learn: 6528.8389411	total: 1m 7s	remaining: 1m 5s
2538:	learn: 6528.6042262	total: 1m 7s	remaining: 1m 5s
2539:	learn: 6525.4895434	total: 1m 7s	remaining: 1m 5s
2540:	learn: 6523.8006736	total: 1m 7s	remaining: 1m 5s
2541:	learn: 6519.8922486	total: 1m 7s	remaining: 1m 5s
2542:	learn: 6518.4991565	total: 1m 7s	remaining: 1m 5s
2543:	learn: 6515.0503728	total: 1m 7s	remaining: 1m 5s
2544:	learn: 6514.1660765	total: 1m 7s	remaining: 1m 4s
2545:	learn: 6512.3206195	total: 1m 7s	remaining: 1m 4s
2546:	learn: 6512.1566352	total: 1m 7s	remaining: 1m 4s
2547:	learn: 6510.4568349	total: 1m 7s	remaining: 1m 4s
2548:	learn: 6510.2037628	total: 1m 7s	remaining: 1m 4s
2549:	learn: 6508.9971614	total: 1m 7s	remaining

2680:	learn: 6257.4120614	total: 1m 11s	remaining: 1m 1s
2681:	learn: 6256.8637666	total: 1m 11s	remaining: 1m 1s
2682:	learn: 6252.9313319	total: 1m 11s	remaining: 1m 1s
2683:	learn: 6252.0553382	total: 1m 11s	remaining: 1m 1s
2684:	learn: 6251.0212955	total: 1m 11s	remaining: 1m 1s
2685:	learn: 6247.8645038	total: 1m 11s	remaining: 1m 1s
2686:	learn: 6245.4669251	total: 1m 11s	remaining: 1m 1s
2687:	learn: 6245.3033893	total: 1m 11s	remaining: 1m 1s
2688:	learn: 6245.2327452	total: 1m 11s	remaining: 1m 1s
2689:	learn: 6242.6024843	total: 1m 11s	remaining: 1m 1s
2690:	learn: 6236.9998832	total: 1m 11s	remaining: 1m 1s
2691:	learn: 6233.8724749	total: 1m 11s	remaining: 1m 1s
2692:	learn: 6233.7987666	total: 1m 11s	remaining: 1m 1s
2693:	learn: 6233.7163267	total: 1m 11s	remaining: 1m 1s
2694:	learn: 6232.9347878	total: 1m 11s	remaining: 1m 1s
2695:	learn: 6231.7021628	total: 1m 11s	remaining: 1m 1s
2696:	learn: 6231.1929559	total: 1m 11s	remaining: 1m 1s
2697:	learn: 6227.5908508	total

2828:	learn: 6016.4878805	total: 1m 15s	remaining: 57.7s
2829:	learn: 6014.0538974	total: 1m 15s	remaining: 57.7s
2830:	learn: 6011.7571631	total: 1m 15s	remaining: 57.7s
2831:	learn: 6010.3089935	total: 1m 15s	remaining: 57.6s
2832:	learn: 6009.1498486	total: 1m 15s	remaining: 57.6s
2833:	learn: 6006.4887427	total: 1m 15s	remaining: 57.6s
2834:	learn: 6005.7176018	total: 1m 15s	remaining: 57.5s
2835:	learn: 6005.6553116	total: 1m 15s	remaining: 57.5s
2836:	learn: 6005.0833506	total: 1m 15s	remaining: 57.5s
2837:	learn: 6003.0254360	total: 1m 15s	remaining: 57.5s
2838:	learn: 5997.2581203	total: 1m 15s	remaining: 57.4s
2839:	learn: 5994.9582845	total: 1m 15s	remaining: 57.4s
2840:	learn: 5994.7327259	total: 1m 15s	remaining: 57.4s
2841:	learn: 5992.9689523	total: 1m 15s	remaining: 57.4s
2842:	learn: 5992.9526285	total: 1m 15s	remaining: 57.3s
2843:	learn: 5990.1422653	total: 1m 15s	remaining: 57.3s
2844:	learn: 5987.0428486	total: 1m 15s	remaining: 57.3s
2845:	learn: 5986.7368487	total

2978:	learn: 5791.2907315	total: 1m 19s	remaining: 54s
2979:	learn: 5790.8258739	total: 1m 19s	remaining: 54s
2980:	learn: 5789.7602335	total: 1m 19s	remaining: 53.9s
2981:	learn: 5789.7399447	total: 1m 19s	remaining: 53.9s
2982:	learn: 5788.4703740	total: 1m 19s	remaining: 53.9s
2983:	learn: 5788.0875946	total: 1m 19s	remaining: 53.9s
2984:	learn: 5787.1416627	total: 1m 19s	remaining: 53.8s
2985:	learn: 5786.0385042	total: 1m 19s	remaining: 53.8s
2986:	learn: 5785.8535224	total: 1m 19s	remaining: 53.8s
2987:	learn: 5785.1996086	total: 1m 19s	remaining: 53.7s
2988:	learn: 5782.5223447	total: 1m 19s	remaining: 53.7s
2989:	learn: 5781.8880854	total: 1m 19s	remaining: 53.7s
2990:	learn: 5780.4277514	total: 1m 19s	remaining: 53.7s
2991:	learn: 5780.3627235	total: 1m 19s	remaining: 53.6s
2992:	learn: 5777.6735235	total: 1m 19s	remaining: 53.6s
2993:	learn: 5776.5268872	total: 1m 19s	remaining: 53.6s
2994:	learn: 5773.5563683	total: 1m 19s	remaining: 53.5s
2995:	learn: 5769.9964681	total: 1m

3125:	learn: 5611.8224590	total: 1m 23s	remaining: 50.3s
3126:	learn: 5610.5164799	total: 1m 23s	remaining: 50.2s
3127:	learn: 5607.8997943	total: 1m 23s	remaining: 50.2s
3128:	learn: 5607.4083857	total: 1m 23s	remaining: 50.2s
3129:	learn: 5607.3190170	total: 1m 23s	remaining: 50.1s
3130:	learn: 5604.6350267	total: 1m 23s	remaining: 50.1s
3131:	learn: 5599.9348802	total: 1m 23s	remaining: 50.1s
3132:	learn: 5598.9889304	total: 1m 24s	remaining: 50.1s
3133:	learn: 5597.8445787	total: 1m 24s	remaining: 50s
3134:	learn: 5595.7955826	total: 1m 24s	remaining: 50s
3135:	learn: 5593.5381669	total: 1m 24s	remaining: 50s
3136:	learn: 5592.9029555	total: 1m 24s	remaining: 49.9s
3137:	learn: 5592.6530570	total: 1m 24s	remaining: 49.9s
3138:	learn: 5591.1816978	total: 1m 24s	remaining: 49.9s
3139:	learn: 5589.5875161	total: 1m 24s	remaining: 49.9s
3140:	learn: 5589.4273864	total: 1m 24s	remaining: 49.8s
3141:	learn: 5587.7411778	total: 1m 24s	remaining: 49.8s
3142:	learn: 5586.1406792	total: 1m 2

3273:	learn: 5422.2009341	total: 1m 28s	remaining: 46.4s
3274:	learn: 5419.4414285	total: 1m 28s	remaining: 46.4s
3275:	learn: 5418.8292222	total: 1m 28s	remaining: 46.4s
3276:	learn: 5416.7586076	total: 1m 28s	remaining: 46.4s
3277:	learn: 5415.7197036	total: 1m 28s	remaining: 46.3s
3278:	learn: 5412.4523365	total: 1m 28s	remaining: 46.3s
3279:	learn: 5409.7744640	total: 1m 28s	remaining: 46.3s
3280:	learn: 5409.7704885	total: 1m 28s	remaining: 46.2s
3281:	learn: 5408.7111681	total: 1m 28s	remaining: 46.2s
3282:	learn: 5407.7738120	total: 1m 28s	remaining: 46.2s
3283:	learn: 5406.1409777	total: 1m 28s	remaining: 46.2s
3284:	learn: 5406.0875947	total: 1m 28s	remaining: 46.1s
3285:	learn: 5406.0005814	total: 1m 28s	remaining: 46.1s
3286:	learn: 5400.8912022	total: 1m 28s	remaining: 46.1s
3287:	learn: 5399.8216527	total: 1m 28s	remaining: 46.1s
3288:	learn: 5398.9986623	total: 1m 28s	remaining: 46s
3289:	learn: 5397.7933879	total: 1m 28s	remaining: 46s
3290:	learn: 5393.2592589	total: 1m

3418:	learn: 5240.0016197	total: 1m 32s	remaining: 42.6s
3419:	learn: 5239.6683999	total: 1m 32s	remaining: 42.6s
3420:	learn: 5239.3049825	total: 1m 32s	remaining: 42.5s
3421:	learn: 5236.8134618	total: 1m 32s	remaining: 42.5s
3422:	learn: 5235.2856223	total: 1m 32s	remaining: 42.5s
3423:	learn: 5233.9953323	total: 1m 32s	remaining: 42.5s
3424:	learn: 5233.6602762	total: 1m 32s	remaining: 42.4s
3425:	learn: 5231.4917869	total: 1m 32s	remaining: 42.4s
3426:	learn: 5231.3846691	total: 1m 32s	remaining: 42.4s
3427:	learn: 5229.2387834	total: 1m 32s	remaining: 42.4s
3428:	learn: 5228.0421793	total: 1m 32s	remaining: 42.3s
3429:	learn: 5226.8373404	total: 1m 32s	remaining: 42.3s
3430:	learn: 5226.4701090	total: 1m 32s	remaining: 42.3s
3431:	learn: 5226.2335673	total: 1m 32s	remaining: 42.3s
3432:	learn: 5225.9891385	total: 1m 32s	remaining: 42.3s
3433:	learn: 5224.6842706	total: 1m 32s	remaining: 42.3s
3434:	learn: 5222.7363912	total: 1m 32s	remaining: 42.2s
3435:	learn: 5221.9898821	total

3570:	learn: 5086.8653865	total: 1m 36s	remaining: 38.6s
3571:	learn: 5085.9043410	total: 1m 36s	remaining: 38.6s
3572:	learn: 5085.4619612	total: 1m 36s	remaining: 38.6s
3573:	learn: 5084.5168264	total: 1m 36s	remaining: 38.6s
3574:	learn: 5083.8122373	total: 1m 36s	remaining: 38.5s
3575:	learn: 5083.5196008	total: 1m 36s	remaining: 38.5s
3576:	learn: 5082.4884398	total: 1m 36s	remaining: 38.5s
3577:	learn: 5082.2247674	total: 1m 36s	remaining: 38.4s
3578:	learn: 5081.8163537	total: 1m 36s	remaining: 38.4s
3579:	learn: 5081.4592636	total: 1m 36s	remaining: 38.4s
3580:	learn: 5081.4421273	total: 1m 36s	remaining: 38.4s
3581:	learn: 5081.3161986	total: 1m 36s	remaining: 38.3s
3582:	learn: 5080.5763204	total: 1m 36s	remaining: 38.3s
3583:	learn: 5079.1073043	total: 1m 36s	remaining: 38.3s
3584:	learn: 5076.8812165	total: 1m 36s	remaining: 38.3s
3585:	learn: 5074.6303013	total: 1m 37s	remaining: 38.3s
3586:	learn: 5073.7155901	total: 1m 37s	remaining: 38.2s
3587:	learn: 5073.6366468	total

3718:	learn: 4909.9722292	total: 1m 40s	remaining: 34.7s
3719:	learn: 4908.3792604	total: 1m 40s	remaining: 34.7s
3720:	learn: 4907.4980487	total: 1m 40s	remaining: 34.7s
3721:	learn: 4907.0171589	total: 1m 40s	remaining: 34.6s
3722:	learn: 4906.8503322	total: 1m 40s	remaining: 34.6s
3723:	learn: 4906.7987480	total: 1m 40s	remaining: 34.6s
3724:	learn: 4906.2842685	total: 1m 40s	remaining: 34.5s
3725:	learn: 4905.0669255	total: 1m 40s	remaining: 34.5s
3726:	learn: 4903.7635094	total: 1m 40s	remaining: 34.5s
3727:	learn: 4901.9530211	total: 1m 41s	remaining: 34.5s
3728:	learn: 4900.6419212	total: 1m 41s	remaining: 34.4s
3729:	learn: 4900.5406006	total: 1m 41s	remaining: 34.4s
3730:	learn: 4900.1010998	total: 1m 41s	remaining: 34.4s
3731:	learn: 4899.7158434	total: 1m 41s	remaining: 34.4s
3732:	learn: 4899.5567941	total: 1m 41s	remaining: 34.3s
3733:	learn: 4899.4160413	total: 1m 41s	remaining: 34.3s
3734:	learn: 4898.4027740	total: 1m 41s	remaining: 34.3s
3735:	learn: 4896.0526028	total

3866:	learn: 4760.9230476	total: 1m 44s	remaining: 30.7s
3867:	learn: 4758.8733022	total: 1m 44s	remaining: 30.7s
3868:	learn: 4758.2697497	total: 1m 44s	remaining: 30.7s
3869:	learn: 4757.7755999	total: 1m 44s	remaining: 30.7s
3870:	learn: 4757.1108441	total: 1m 45s	remaining: 30.6s
3871:	learn: 4756.9816059	total: 1m 45s	remaining: 30.6s
3872:	learn: 4756.2401179	total: 1m 45s	remaining: 30.6s
3873:	learn: 4754.1580874	total: 1m 45s	remaining: 30.5s
3874:	learn: 4753.0147342	total: 1m 45s	remaining: 30.5s
3875:	learn: 4752.0932164	total: 1m 45s	remaining: 30.5s
3876:	learn: 4750.4144342	total: 1m 45s	remaining: 30.5s
3877:	learn: 4748.8739005	total: 1m 45s	remaining: 30.4s
3878:	learn: 4744.8444360	total: 1m 45s	remaining: 30.4s
3879:	learn: 4742.4582715	total: 1m 45s	remaining: 30.4s
3880:	learn: 4740.8730072	total: 1m 45s	remaining: 30.4s
3881:	learn: 4739.5019042	total: 1m 45s	remaining: 30.3s
3882:	learn: 4737.7151288	total: 1m 45s	remaining: 30.3s
3883:	learn: 4736.3123762	total

4013:	learn: 4610.8159472	total: 1m 49s	remaining: 26.9s
4014:	learn: 4610.5128376	total: 1m 49s	remaining: 26.9s
4015:	learn: 4610.4101482	total: 1m 49s	remaining: 26.8s
4016:	learn: 4608.5899422	total: 1m 49s	remaining: 26.8s
4017:	learn: 4607.3882319	total: 1m 49s	remaining: 26.8s
4018:	learn: 4606.6597555	total: 1m 49s	remaining: 26.8s
4019:	learn: 4606.6222323	total: 1m 49s	remaining: 26.7s
4020:	learn: 4605.7308875	total: 1m 49s	remaining: 26.7s
4021:	learn: 4603.9564308	total: 1m 49s	remaining: 26.7s
4022:	learn: 4603.7398888	total: 1m 49s	remaining: 26.6s
4023:	learn: 4602.0869526	total: 1m 49s	remaining: 26.6s
4024:	learn: 4601.7349466	total: 1m 49s	remaining: 26.6s
4025:	learn: 4600.9144787	total: 1m 49s	remaining: 26.6s
4026:	learn: 4597.9416985	total: 1m 49s	remaining: 26.5s
4027:	learn: 4595.7642431	total: 1m 49s	remaining: 26.5s
4028:	learn: 4592.5346644	total: 1m 49s	remaining: 26.5s
4029:	learn: 4591.7133548	total: 1m 49s	remaining: 26.5s
4030:	learn: 4591.6903043	total

4166:	learn: 4454.1393751	total: 1m 53s	remaining: 22.8s
4167:	learn: 4454.0404978	total: 1m 53s	remaining: 22.7s
4168:	learn: 4453.8163847	total: 1m 53s	remaining: 22.7s
4169:	learn: 4451.4082338	total: 1m 53s	remaining: 22.7s
4170:	learn: 4448.7829045	total: 1m 53s	remaining: 22.6s
4171:	learn: 4448.5219919	total: 1m 53s	remaining: 22.6s
4172:	learn: 4447.8953103	total: 1m 54s	remaining: 22.6s
4173:	learn: 4447.7918409	total: 1m 54s	remaining: 22.6s
4174:	learn: 4447.4195882	total: 1m 54s	remaining: 22.6s
4175:	learn: 4447.2169858	total: 1m 54s	remaining: 22.5s
4176:	learn: 4447.0247135	total: 1m 54s	remaining: 22.5s
4177:	learn: 4446.7931657	total: 1m 54s	remaining: 22.5s
4178:	learn: 4444.9115968	total: 1m 54s	remaining: 22.4s
4179:	learn: 4444.7816693	total: 1m 54s	remaining: 22.4s
4180:	learn: 4444.0659203	total: 1m 54s	remaining: 22.4s
4181:	learn: 4441.7072138	total: 1m 54s	remaining: 22.4s
4182:	learn: 4441.5565620	total: 1m 54s	remaining: 22.3s
4183:	learn: 4440.7436550	total

4315:	learn: 4324.8682250	total: 1m 58s	remaining: 18.7s
4316:	learn: 4324.4354878	total: 1m 58s	remaining: 18.7s
4317:	learn: 4324.1578675	total: 1m 58s	remaining: 18.7s
4318:	learn: 4322.2672255	total: 1m 58s	remaining: 18.6s
4319:	learn: 4322.1683668	total: 1m 58s	remaining: 18.6s
4320:	learn: 4321.9017164	total: 1m 58s	remaining: 18.6s
4321:	learn: 4321.4035650	total: 1m 58s	remaining: 18.6s
4322:	learn: 4321.0806460	total: 1m 58s	remaining: 18.5s
4323:	learn: 4320.8030434	total: 1m 58s	remaining: 18.5s
4324:	learn: 4320.2061630	total: 1m 58s	remaining: 18.5s
4325:	learn: 4319.2866531	total: 1m 58s	remaining: 18.5s
4326:	learn: 4319.1881376	total: 1m 58s	remaining: 18.4s
4327:	learn: 4318.7525563	total: 1m 58s	remaining: 18.4s
4328:	learn: 4317.1256977	total: 1m 58s	remaining: 18.4s
4329:	learn: 4316.9108775	total: 1m 58s	remaining: 18.3s
4330:	learn: 4315.4058159	total: 1m 58s	remaining: 18.3s
4331:	learn: 4314.7447719	total: 1m 58s	remaining: 18.3s
4332:	learn: 4314.5104644	total

4466:	learn: 4198.4417481	total: 2m 2s	remaining: 14.6s
4467:	learn: 4197.3632304	total: 2m 2s	remaining: 14.6s
4468:	learn: 4195.3533492	total: 2m 2s	remaining: 14.6s
4469:	learn: 4195.0596914	total: 2m 2s	remaining: 14.6s
4470:	learn: 4194.9755013	total: 2m 2s	remaining: 14.5s
4471:	learn: 4194.9596582	total: 2m 2s	remaining: 14.5s
4472:	learn: 4194.9367658	total: 2m 2s	remaining: 14.5s
4473:	learn: 4194.2584691	total: 2m 2s	remaining: 14.4s
4474:	learn: 4194.2411108	total: 2m 2s	remaining: 14.4s
4475:	learn: 4192.9999324	total: 2m 2s	remaining: 14.4s
4476:	learn: 4191.5487042	total: 2m 2s	remaining: 14.4s
4477:	learn: 4190.5409855	total: 2m 2s	remaining: 14.3s
4478:	learn: 4189.4454880	total: 2m 2s	remaining: 14.3s
4479:	learn: 4188.9771124	total: 2m 2s	remaining: 14.3s
4480:	learn: 4188.9381664	total: 2m 2s	remaining: 14.2s
4481:	learn: 4188.9323535	total: 2m 3s	remaining: 14.2s
4482:	learn: 4185.9993757	total: 2m 3s	remaining: 14.2s
4483:	learn: 4184.9171731	total: 2m 3s	remaining

4617:	learn: 4079.5052446	total: 2m 7s	remaining: 10.5s
4618:	learn: 4079.4764607	total: 2m 7s	remaining: 10.5s
4619:	learn: 4077.2333630	total: 2m 7s	remaining: 10.5s
4620:	learn: 4075.3151022	total: 2m 7s	remaining: 10.5s
4621:	learn: 4075.2315302	total: 2m 7s	remaining: 10.4s
4622:	learn: 4074.6300652	total: 2m 7s	remaining: 10.4s
4623:	learn: 4073.1033087	total: 2m 7s	remaining: 10.4s
4624:	learn: 4072.8467646	total: 2m 7s	remaining: 10.3s
4625:	learn: 4071.4883737	total: 2m 7s	remaining: 10.3s
4626:	learn: 4070.8727453	total: 2m 7s	remaining: 10.3s
4627:	learn: 4070.0309684	total: 2m 7s	remaining: 10.3s
4628:	learn: 4069.9425209	total: 2m 7s	remaining: 10.2s
4629:	learn: 4067.7401506	total: 2m 7s	remaining: 10.2s
4630:	learn: 4067.2423793	total: 2m 7s	remaining: 10.2s
4631:	learn: 4066.7655260	total: 2m 7s	remaining: 10.2s
4632:	learn: 4066.3316063	total: 2m 7s	remaining: 10.1s
4633:	learn: 4065.9617194	total: 2m 7s	remaining: 10.1s
4634:	learn: 4065.7393414	total: 2m 7s	remaining

4766:	learn: 3959.9640289	total: 2m 11s	remaining: 6.44s
4767:	learn: 3958.5750281	total: 2m 11s	remaining: 6.41s
4768:	learn: 3958.1732288	total: 2m 11s	remaining: 6.38s
4769:	learn: 3958.1051980	total: 2m 11s	remaining: 6.36s
4770:	learn: 3958.0765205	total: 2m 11s	remaining: 6.33s
4771:	learn: 3957.0859457	total: 2m 11s	remaining: 6.3s
4772:	learn: 3956.4137590	total: 2m 11s	remaining: 6.27s
4773:	learn: 3955.4411773	total: 2m 11s	remaining: 6.25s
4774:	learn: 3955.3775901	total: 2m 11s	remaining: 6.22s
4775:	learn: 3954.1466921	total: 2m 11s	remaining: 6.19s
4776:	learn: 3953.9834357	total: 2m 12s	remaining: 6.16s
4777:	learn: 3953.4374606	total: 2m 12s	remaining: 6.13s
4778:	learn: 3952.9317464	total: 2m 12s	remaining: 6.11s
4779:	learn: 3951.2217657	total: 2m 12s	remaining: 6.08s
4780:	learn: 3950.9208595	total: 2m 12s	remaining: 6.05s
4781:	learn: 3950.4919067	total: 2m 12s	remaining: 6.02s
4782:	learn: 3949.0859632	total: 2m 12s	remaining: 6s
4783:	learn: 3948.1764208	total: 2m

4911:	learn: 3853.3886746	total: 2m 15s	remaining: 2.43s
4912:	learn: 3853.1835164	total: 2m 15s	remaining: 2.41s
4913:	learn: 3852.9499936	total: 2m 15s	remaining: 2.38s
4914:	learn: 3850.7786251	total: 2m 15s	remaining: 2.35s
4915:	learn: 3850.7687870	total: 2m 15s	remaining: 2.32s
4916:	learn: 3848.8716067	total: 2m 16s	remaining: 2.29s
4917:	learn: 3848.4122602	total: 2m 16s	remaining: 2.27s
4918:	learn: 3848.0100708	total: 2m 16s	remaining: 2.24s
4919:	learn: 3847.2701023	total: 2m 16s	remaining: 2.21s
4920:	learn: 3847.1575472	total: 2m 16s	remaining: 2.19s
4921:	learn: 3846.1659171	total: 2m 16s	remaining: 2.16s
4922:	learn: 3845.9047647	total: 2m 16s	remaining: 2.13s
4923:	learn: 3843.4287512	total: 2m 16s	remaining: 2.1s
4924:	learn: 3842.9020162	total: 2m 16s	remaining: 2.08s
4925:	learn: 3842.8211314	total: 2m 16s	remaining: 2.05s
4926:	learn: 3842.2928608	total: 2m 16s	remaining: 2.02s
4927:	learn: 3842.1090204	total: 2m 16s	remaining: 1.99s
4928:	learn: 3841.3057953	total:

GridSearchCV(cv=2,
             estimator=<catboost.core.CatBoostRegressor object at 0x0000024705ED2DC0>,
             n_jobs=-1,
             param_grid={'depth': [2, 4, 6], 'iterations': [1000, 2000, 5000],
                         'learning_rate': [0.01, 0.05, 0.1]})

--- 2877.3895540237427 seconds ---
 Results from Grid Search 

 The best estimator across ALL searched params:

 The best score across ALL searched params:
 0.8850863304250196

 The best parameters across ALL searched params:
 {'depth': 4, 'iterations': 5000, 'learning_rate': 0.05}


## After a mulititude of different combination of three parameters  (they were the only parameters I could use on my silly little laptop before it croaks, I had to use a DIY external fan to cool it during the gridsearchcv running) I was able to slowly change them depending on the output everytime until I came up with the parameters in the model below. so i mixed in GridSearchCV and some manual parameter tuning to make use of my limited resources so my model wont be the best but it will be good enough for this instance. GridSearchCV gave me 5000 iterations, a learning rate of 0.05 and a depth of 4. the best score was 0.8850863304250196 and it took 2877.3895540237427 seconds to run the GridSearchCV for the above parameters.

In [26]:
cat_model = CatBoostRegressor(loss_function='RMSE',depth= 4,iterations = 1838,learning_rate = 0.05)
cat_model.fit( X_train, Y_train, cat_features = cat_features, eval_set=(X_test, Y_test), plot=True)

MetricVisualizer(layout=Layout(align_self='stretch', height='500px'))

0:	learn: 78273.0604381	test: 71384.2981383	best: 71384.2981383 (0)	total: 27.3ms	remaining: 50.2s
1:	learn: 75911.6045995	test: 69217.9089604	best: 69217.9089604 (1)	total: 57.7ms	remaining: 53s
2:	learn: 73658.2964609	test: 67122.9583159	best: 67122.9583159 (2)	total: 81.1ms	remaining: 49.6s
3:	learn: 71551.0364917	test: 65313.9611690	best: 65313.9611690 (3)	total: 116ms	remaining: 53s
4:	learn: 69672.5930756	test: 63539.4799175	best: 63539.4799175 (4)	total: 140ms	remaining: 51.3s
5:	learn: 67682.7492828	test: 61816.8375984	best: 61816.8375984 (5)	total: 164ms	remaining: 50.1s
6:	learn: 65676.5117176	test: 59869.1191163	best: 59869.1191163 (6)	total: 189ms	remaining: 49.3s
7:	learn: 64084.2006646	test: 58564.7329773	best: 58564.7329773 (7)	total: 225ms	remaining: 51.5s
8:	learn: 62245.5981248	test: 56746.3667492	best: 56746.3667492 (8)	total: 263ms	remaining: 53.5s
9:	learn: 60557.2218109	test: 55212.3154818	best: 55212.3154818 (9)	total: 294ms	remaining: 53.7s
10:	learn: 59211.3438

83:	learn: 25481.0713208	test: 30360.9804767	best: 30360.9804767 (83)	total: 2.19s	remaining: 45.7s
84:	learn: 25329.8387777	test: 30323.1209652	best: 30323.1209652 (84)	total: 2.22s	remaining: 45.7s
85:	learn: 25257.6803784	test: 30275.3665058	best: 30275.3665058 (85)	total: 2.25s	remaining: 45.8s
86:	learn: 25222.4161537	test: 30293.3608300	best: 30275.3665058 (85)	total: 2.27s	remaining: 45.7s
87:	learn: 25145.1071827	test: 30234.7535795	best: 30234.7535795 (87)	total: 2.29s	remaining: 45.6s
88:	learn: 25091.7822542	test: 30230.7036671	best: 30230.7036671 (88)	total: 2.33s	remaining: 45.7s
89:	learn: 24996.5905050	test: 30211.8939802	best: 30211.8939802 (89)	total: 2.35s	remaining: 45.7s
90:	learn: 24972.6556547	test: 30211.9131540	best: 30211.8939802 (89)	total: 2.37s	remaining: 45.4s
91:	learn: 24878.2939326	test: 30194.4420474	best: 30194.4420474 (91)	total: 2.39s	remaining: 45.4s
92:	learn: 24772.4930145	test: 30179.8784754	best: 30179.8784754 (92)	total: 2.42s	remaining: 45.4s


165:	learn: 20565.1052771	test: 28463.3166727	best: 28403.4530950 (141)	total: 4.38s	remaining: 44.1s
166:	learn: 20549.4927628	test: 28494.1891510	best: 28403.4530950 (141)	total: 4.41s	remaining: 44.1s
167:	learn: 20534.8130606	test: 28525.8010758	best: 28403.4530950 (141)	total: 4.43s	remaining: 44.1s
168:	learn: 20470.1832750	test: 28437.1013685	best: 28403.4530950 (141)	total: 4.46s	remaining: 44.1s
169:	learn: 20456.5685781	test: 28467.7251591	best: 28403.4530950 (141)	total: 4.49s	remaining: 44s
170:	learn: 20417.8577176	test: 28459.5986387	best: 28403.4530950 (141)	total: 4.51s	remaining: 44s
171:	learn: 20400.7154555	test: 28449.6574966	best: 28403.4530950 (141)	total: 4.53s	remaining: 43.9s
172:	learn: 20387.8045687	test: 28477.9988982	best: 28403.4530950 (141)	total: 4.55s	remaining: 43.8s
173:	learn: 20335.9534696	test: 28415.3495207	best: 28403.4530950 (141)	total: 4.58s	remaining: 43.8s
174:	learn: 20291.8175362	test: 28357.8798838	best: 28357.8798838 (174)	total: 4.6s	re

254:	learn: 18615.5802952	test: 27852.2004351	best: 27852.2004351 (254)	total: 6.71s	remaining: 41.7s
255:	learn: 18610.1120840	test: 27874.0612633	best: 27852.2004351 (254)	total: 6.74s	remaining: 41.7s
256:	learn: 18584.7717695	test: 27849.6302078	best: 27849.6302078 (256)	total: 6.76s	remaining: 41.6s
257:	learn: 18551.2429184	test: 27772.9166425	best: 27772.9166425 (257)	total: 6.78s	remaining: 41.5s
258:	learn: 18496.8765669	test: 27762.5075618	best: 27762.5075618 (258)	total: 6.82s	remaining: 41.6s
259:	learn: 18468.3589518	test: 27701.7571837	best: 27701.7571837 (259)	total: 6.84s	remaining: 41.5s
260:	learn: 18462.2424908	test: 27720.7351364	best: 27701.7571837 (259)	total: 6.86s	remaining: 41.5s
261:	learn: 18455.9542550	test: 27739.3221511	best: 27701.7571837 (259)	total: 6.89s	remaining: 41.4s
262:	learn: 18446.9207833	test: 27719.7833184	best: 27701.7571837 (259)	total: 6.91s	remaining: 41.4s
263:	learn: 18433.3683225	test: 27723.5299359	best: 27701.7571837 (259)	total: 6.9

341:	learn: 17087.4688623	test: 27270.9431186	best: 27252.4421561 (333)	total: 8.89s	remaining: 38.9s
342:	learn: 17070.6519402	test: 27249.4414270	best: 27249.4414270 (342)	total: 8.91s	remaining: 38.8s
343:	learn: 17065.0384929	test: 27248.5746385	best: 27248.5746385 (343)	total: 8.93s	remaining: 38.8s
344:	learn: 17063.6862108	test: 27251.6840629	best: 27248.5746385 (343)	total: 8.96s	remaining: 38.8s
345:	learn: 17044.8135904	test: 27265.1925792	best: 27248.5746385 (343)	total: 8.98s	remaining: 38.7s
346:	learn: 17010.6972438	test: 27247.8373809	best: 27247.8373809 (346)	total: 9.02s	remaining: 38.8s
347:	learn: 17000.3703234	test: 27269.8309109	best: 27247.8373809 (346)	total: 9.04s	remaining: 38.7s
348:	learn: 16979.6120841	test: 27254.1591441	best: 27247.8373809 (346)	total: 9.07s	remaining: 38.7s
349:	learn: 16978.7149835	test: 27255.3652940	best: 27247.8373809 (346)	total: 9.1s	remaining: 38.7s
350:	learn: 16956.7389509	test: 27269.4217643	best: 27247.8373809 (346)	total: 9.12

425:	learn: 15787.0899417	test: 26878.2467943	best: 26877.9137960 (420)	total: 11.1s	remaining: 36.6s
426:	learn: 15775.6609438	test: 26859.6537508	best: 26859.6537508 (426)	total: 11.1s	remaining: 36.6s
427:	learn: 15759.4955962	test: 26855.9639814	best: 26855.9639814 (427)	total: 11.1s	remaining: 36.6s
428:	learn: 15737.3553534	test: 26856.3356944	best: 26855.9639814 (427)	total: 11.1s	remaining: 36.6s
429:	learn: 15721.1841210	test: 26857.1212734	best: 26855.9639814 (427)	total: 11.2s	remaining: 36.6s
430:	learn: 15706.8292372	test: 26846.0823565	best: 26846.0823565 (430)	total: 11.2s	remaining: 36.6s
431:	learn: 15674.4771610	test: 26872.8191971	best: 26846.0823565 (430)	total: 11.2s	remaining: 36.6s
432:	learn: 15671.5676076	test: 26874.1959706	best: 26846.0823565 (430)	total: 11.3s	remaining: 36.6s
433:	learn: 15666.9795011	test: 26872.4920603	best: 26846.0823565 (430)	total: 11.3s	remaining: 36.5s
434:	learn: 15650.9116638	test: 26870.3269990	best: 26846.0823565 (430)	total: 11.

512:	learn: 14748.8392449	test: 26725.2834567	best: 26698.2190766 (489)	total: 13.4s	remaining: 34.7s
513:	learn: 14745.5165799	test: 26709.8914748	best: 26698.2190766 (489)	total: 13.5s	remaining: 34.7s
514:	learn: 14744.8602760	test: 26709.8904835	best: 26698.2190766 (489)	total: 13.5s	remaining: 34.7s
515:	learn: 14739.6466673	test: 26694.6942949	best: 26694.6942949 (515)	total: 13.5s	remaining: 34.7s
516:	learn: 14736.7075421	test: 26683.1261123	best: 26683.1261123 (516)	total: 13.6s	remaining: 34.7s
517:	learn: 14733.2181188	test: 26686.9225538	best: 26683.1261123 (516)	total: 13.6s	remaining: 34.7s
518:	learn: 14720.4902859	test: 26710.5450768	best: 26683.1261123 (516)	total: 13.6s	remaining: 34.6s
519:	learn: 14719.2753374	test: 26711.7026859	best: 26683.1261123 (516)	total: 13.7s	remaining: 34.6s
520:	learn: 14716.3048280	test: 26711.9259817	best: 26683.1261123 (516)	total: 13.7s	remaining: 34.6s
521:	learn: 14707.7238503	test: 26699.8133913	best: 26683.1261123 (516)	total: 13.

595:	learn: 14055.8261758	test: 26716.4443806	best: 26614.3652971 (565)	total: 15.8s	remaining: 33s
596:	learn: 14038.5322940	test: 26717.9425521	best: 26614.3652971 (565)	total: 15.8s	remaining: 32.9s
597:	learn: 14036.0462830	test: 26720.1448145	best: 26614.3652971 (565)	total: 15.9s	remaining: 32.9s
598:	learn: 14033.6777531	test: 26706.8845173	best: 26614.3652971 (565)	total: 15.9s	remaining: 32.9s
599:	learn: 14017.5865388	test: 26726.4495807	best: 26614.3652971 (565)	total: 15.9s	remaining: 32.8s
600:	learn: 14015.7462728	test: 26732.1717160	best: 26614.3652971 (565)	total: 15.9s	remaining: 32.8s
601:	learn: 14013.2229553	test: 26719.3830348	best: 26614.3652971 (565)	total: 16s	remaining: 32.8s
602:	learn: 14003.9362482	test: 26719.5545846	best: 26614.3652971 (565)	total: 16s	remaining: 32.7s
603:	learn: 13996.5633322	test: 26733.1082932	best: 26614.3652971 (565)	total: 16s	remaining: 32.7s
604:	learn: 13986.4780727	test: 26737.9267361	best: 26614.3652971 (565)	total: 16s	remaini

681:	learn: 13316.0812120	test: 26665.8820761	best: 26614.3652971 (565)	total: 18s	remaining: 30.5s
682:	learn: 13315.3079135	test: 26665.9057999	best: 26614.3652971 (565)	total: 18s	remaining: 30.5s
683:	learn: 13309.6492940	test: 26671.4189267	best: 26614.3652971 (565)	total: 18.1s	remaining: 30.5s
684:	learn: 13301.2698642	test: 26674.1309284	best: 26614.3652971 (565)	total: 18.1s	remaining: 30.4s
685:	learn: 13300.7260662	test: 26667.5077908	best: 26614.3652971 (565)	total: 18.1s	remaining: 30.4s
686:	learn: 13293.5221812	test: 26653.7803494	best: 26614.3652971 (565)	total: 18.1s	remaining: 30.4s
687:	learn: 13292.1201822	test: 26647.5949331	best: 26614.3652971 (565)	total: 18.1s	remaining: 30.3s
688:	learn: 13291.6231207	test: 26641.3356945	best: 26614.3652971 (565)	total: 18.2s	remaining: 30.3s
689:	learn: 13288.0597308	test: 26645.4512486	best: 26614.3652971 (565)	total: 18.2s	remaining: 30.3s
690:	learn: 13275.3095588	test: 26643.5570031	best: 26614.3652971 (565)	total: 18.2s	r

765:	learn: 12783.9705767	test: 26563.1177657	best: 26545.7931075 (751)	total: 20.2s	remaining: 28.3s
766:	learn: 12776.0856966	test: 26567.5868546	best: 26545.7931075 (751)	total: 20.3s	remaining: 28.3s
767:	learn: 12768.0134082	test: 26571.8976665	best: 26545.7931075 (751)	total: 20.3s	remaining: 28.3s
768:	learn: 12767.2188515	test: 26575.5799673	best: 26545.7931075 (751)	total: 20.3s	remaining: 28.2s
769:	learn: 12763.2929827	test: 26580.5128532	best: 26545.7931075 (751)	total: 20.3s	remaining: 28.2s
770:	learn: 12762.9132005	test: 26575.3203165	best: 26545.7931075 (751)	total: 20.4s	remaining: 28.2s
771:	learn: 12755.4802333	test: 26579.1747797	best: 26545.7931075 (751)	total: 20.4s	remaining: 28.2s
772:	learn: 12753.4200014	test: 26578.0288766	best: 26545.7931075 (751)	total: 20.4s	remaining: 28.1s
773:	learn: 12747.5877827	test: 26577.1766238	best: 26545.7931075 (751)	total: 20.4s	remaining: 28.1s
774:	learn: 12735.4105592	test: 26582.5260132	best: 26545.7931075 (751)	total: 20.

848:	learn: 12273.3123608	test: 26689.2395522	best: 26545.7931075 (751)	total: 22.4s	remaining: 26.1s
849:	learn: 12267.3347428	test: 26694.0221424	best: 26545.7931075 (751)	total: 22.5s	remaining: 26.1s
850:	learn: 12260.9793463	test: 26691.7006445	best: 26545.7931075 (751)	total: 22.5s	remaining: 26.1s
851:	learn: 12256.2565867	test: 26687.5626553	best: 26545.7931075 (751)	total: 22.6s	remaining: 26.1s
852:	learn: 12252.5068287	test: 26686.8491225	best: 26545.7931075 (751)	total: 22.6s	remaining: 26.1s
853:	learn: 12250.3064470	test: 26686.8878458	best: 26545.7931075 (751)	total: 22.6s	remaining: 26.1s
854:	learn: 12246.2946060	test: 26684.1881345	best: 26545.7931075 (751)	total: 22.7s	remaining: 26.1s
855:	learn: 12242.5766886	test: 26682.3299762	best: 26545.7931075 (751)	total: 22.7s	remaining: 26.1s
856:	learn: 12235.4588599	test: 26672.0111098	best: 26545.7931075 (751)	total: 22.8s	remaining: 26.1s
857:	learn: 12218.4676336	test: 26668.3936652	best: 26545.7931075 (751)	total: 22.

930:	learn: 11725.0496162	test: 26688.8469516	best: 26545.7931075 (751)	total: 24.9s	remaining: 24.2s
931:	learn: 11715.9074165	test: 26707.5025617	best: 26545.7931075 (751)	total: 24.9s	remaining: 24.2s
932:	learn: 11704.2421674	test: 26705.7948720	best: 26545.7931075 (751)	total: 24.9s	remaining: 24.2s
933:	learn: 11700.8741227	test: 26703.6291983	best: 26545.7931075 (751)	total: 25s	remaining: 24.2s
934:	learn: 11685.7851299	test: 26709.3177105	best: 26545.7931075 (751)	total: 25s	remaining: 24.1s
935:	learn: 11680.6962260	test: 26708.8430046	best: 26545.7931075 (751)	total: 25s	remaining: 24.1s
936:	learn: 11677.6934620	test: 26701.4565892	best: 26545.7931075 (751)	total: 25s	remaining: 24.1s
937:	learn: 11669.7441155	test: 26699.7762511	best: 26545.7931075 (751)	total: 25.1s	remaining: 24s
938:	learn: 11664.7456765	test: 26698.5849363	best: 26545.7931075 (751)	total: 25.1s	remaining: 24s
939:	learn: 11661.6316301	test: 26693.2598722	best: 26545.7931075 (751)	total: 25.1s	remaining

1016:	learn: 11183.4788013	test: 26626.3558264	best: 26545.7931075 (751)	total: 27.3s	remaining: 22s
1017:	learn: 11167.5343375	test: 26626.1827565	best: 26545.7931075 (751)	total: 27.3s	remaining: 22s
1018:	learn: 11164.6221534	test: 26630.8981289	best: 26545.7931075 (751)	total: 27.3s	remaining: 22s
1019:	learn: 11160.8691834	test: 26634.4439748	best: 26545.7931075 (751)	total: 27.4s	remaining: 21.9s
1020:	learn: 11154.4691106	test: 26642.1492247	best: 26545.7931075 (751)	total: 27.4s	remaining: 21.9s
1021:	learn: 11149.8292084	test: 26644.1644601	best: 26545.7931075 (751)	total: 27.4s	remaining: 21.9s
1022:	learn: 11137.8734537	test: 26646.8459888	best: 26545.7931075 (751)	total: 27.4s	remaining: 21.8s
1023:	learn: 11136.9487163	test: 26646.9995045	best: 26545.7931075 (751)	total: 27.4s	remaining: 21.8s
1024:	learn: 11133.7592974	test: 26646.2920070	best: 26545.7931075 (751)	total: 27.5s	remaining: 21.8s
1025:	learn: 11121.3304138	test: 26637.3758288	best: 26545.7931075 (751)	total:

1101:	learn: 10789.3547398	test: 26612.5254761	best: 26545.7931075 (751)	total: 29.7s	remaining: 19.8s
1102:	learn: 10788.4808021	test: 26612.4915178	best: 26545.7931075 (751)	total: 29.8s	remaining: 19.8s
1103:	learn: 10782.3003988	test: 26615.5146288	best: 26545.7931075 (751)	total: 29.8s	remaining: 19.8s
1104:	learn: 10760.6553041	test: 26604.6906023	best: 26545.7931075 (751)	total: 29.8s	remaining: 19.8s
1105:	learn: 10747.9718861	test: 26610.3917739	best: 26545.7931075 (751)	total: 29.8s	remaining: 19.7s
1106:	learn: 10746.0071484	test: 26604.5559349	best: 26545.7931075 (751)	total: 29.9s	remaining: 19.7s
1107:	learn: 10745.4127734	test: 26599.7569255	best: 26545.7931075 (751)	total: 29.9s	remaining: 19.7s
1108:	learn: 10741.7589500	test: 26596.0546765	best: 26545.7931075 (751)	total: 29.9s	remaining: 19.7s
1109:	learn: 10737.9486048	test: 26595.3637914	best: 26545.7931075 (751)	total: 30s	remaining: 19.6s
1110:	learn: 10736.6799893	test: 26595.6410972	best: 26545.7931075 (751)	to

1183:	learn: 10374.7479561	test: 26565.9952003	best: 26543.3712136 (1176)	total: 31.9s	remaining: 17.6s
1184:	learn: 10370.5924515	test: 26568.2191733	best: 26543.3712136 (1176)	total: 31.9s	remaining: 17.6s
1185:	learn: 10368.8058667	test: 26570.0479089	best: 26543.3712136 (1176)	total: 32s	remaining: 17.6s
1186:	learn: 10368.2597444	test: 26573.8873939	best: 26543.3712136 (1176)	total: 32s	remaining: 17.5s
1187:	learn: 10368.2531112	test: 26573.8862369	best: 26543.3712136 (1176)	total: 32s	remaining: 17.5s
1188:	learn: 10367.8177562	test: 26569.9651594	best: 26543.3712136 (1176)	total: 32s	remaining: 17.5s
1189:	learn: 10364.6316850	test: 26569.7795230	best: 26543.3712136 (1176)	total: 32.1s	remaining: 17.5s
1190:	learn: 10362.1222561	test: 26570.7905084	best: 26543.3712136 (1176)	total: 32.1s	remaining: 17.4s
1191:	learn: 10360.4420822	test: 26571.0651864	best: 26543.3712136 (1176)	total: 32.1s	remaining: 17.4s
1192:	learn: 10355.2346423	test: 26575.2529929	best: 26543.3712136 (1176

1265:	learn: 10056.9530527	test: 26608.5308157	best: 26543.3712136 (1176)	total: 34.1s	remaining: 15.4s
1266:	learn: 10053.5870940	test: 26607.5412685	best: 26543.3712136 (1176)	total: 34.2s	remaining: 15.4s
1267:	learn: 10043.9806771	test: 26606.8445511	best: 26543.3712136 (1176)	total: 34.2s	remaining: 15.4s
1268:	learn: 10043.6043355	test: 26602.9527421	best: 26543.3712136 (1176)	total: 34.2s	remaining: 15.3s
1269:	learn: 10041.2751503	test: 26602.7350276	best: 26543.3712136 (1176)	total: 34.2s	remaining: 15.3s
1270:	learn: 10037.3087673	test: 26600.7899124	best: 26543.3712136 (1176)	total: 34.2s	remaining: 15.3s
1271:	learn: 10034.8843413	test: 26602.1831447	best: 26543.3712136 (1176)	total: 34.3s	remaining: 15.3s
1272:	learn: 10029.7490094	test: 26603.2669852	best: 26543.3712136 (1176)	total: 34.3s	remaining: 15.2s
1273:	learn: 10026.1860163	test: 26599.8482319	best: 26543.3712136 (1176)	total: 34.3s	remaining: 15.2s
1274:	learn: 10023.1784135	test: 26597.7923996	best: 26543.37121

1348:	learn: 9770.6874838	test: 26544.5474894	best: 26537.4295976 (1344)	total: 36.3s	remaining: 13.1s
1349:	learn: 9770.6871232	test: 26544.5479982	best: 26537.4295976 (1344)	total: 36.3s	remaining: 13.1s
1350:	learn: 9770.0419099	test: 26548.1132449	best: 26537.4295976 (1344)	total: 36.3s	remaining: 13.1s
1351:	learn: 9766.1760352	test: 26548.2458725	best: 26537.4295976 (1344)	total: 36.3s	remaining: 13.1s
1352:	learn: 9762.3207267	test: 26546.9353420	best: 26537.4295976 (1344)	total: 36.4s	remaining: 13s
1353:	learn: 9757.4140517	test: 26549.5129345	best: 26537.4295976 (1344)	total: 36.4s	remaining: 13s
1354:	learn: 9752.3153488	test: 26546.4761139	best: 26537.4295976 (1344)	total: 36.4s	remaining: 13s
1355:	learn: 9752.3047020	test: 26546.4789874	best: 26537.4295976 (1344)	total: 36.4s	remaining: 12.9s
1356:	learn: 9750.3951820	test: 26545.0160975	best: 26537.4295976 (1344)	total: 36.4s	remaining: 12.9s
1357:	learn: 9746.0214471	test: 26544.7853136	best: 26537.4295976 (1344)	total:

1436:	learn: 9488.1668760	test: 26513.8318271	best: 26513.8318271 (1436)	total: 38.7s	remaining: 10.8s
1437:	learn: 9487.6013046	test: 26511.6225841	best: 26511.6225841 (1437)	total: 38.7s	remaining: 10.8s
1438:	learn: 9484.6204375	test: 26503.6208975	best: 26503.6208975 (1438)	total: 38.8s	remaining: 10.7s
1439:	learn: 9483.4604582	test: 26503.7734042	best: 26503.6208975 (1438)	total: 38.8s	remaining: 10.7s
1440:	learn: 9481.5069308	test: 26504.9716114	best: 26503.6208975 (1438)	total: 38.8s	remaining: 10.7s
1441:	learn: 9481.2629194	test: 26505.2823395	best: 26503.6208975 (1438)	total: 38.8s	remaining: 10.7s
1442:	learn: 9480.8187937	test: 26506.3794527	best: 26503.6208975 (1438)	total: 38.9s	remaining: 10.6s
1443:	learn: 9480.5597043	test: 26506.9470755	best: 26503.6208975 (1438)	total: 38.9s	remaining: 10.6s
1444:	learn: 9473.6750361	test: 26503.3263763	best: 26503.3263763 (1444)	total: 38.9s	remaining: 10.6s
1445:	learn: 9465.3075586	test: 26491.6483801	best: 26491.6483801 (1445)	

1523:	learn: 9173.5414229	test: 26524.0534053	best: 26479.7397964 (1447)	total: 41s	remaining: 8.45s
1524:	learn: 9171.8663191	test: 26518.6540928	best: 26479.7397964 (1447)	total: 41.1s	remaining: 8.43s
1525:	learn: 9169.2148776	test: 26521.0852657	best: 26479.7397964 (1447)	total: 41.1s	remaining: 8.4s
1526:	learn: 9163.9378132	test: 26524.0825120	best: 26479.7397964 (1447)	total: 41.1s	remaining: 8.37s
1527:	learn: 9163.3232416	test: 26521.7153964	best: 26479.7397964 (1447)	total: 41.1s	remaining: 8.35s
1528:	learn: 9161.6909650	test: 26522.7366774	best: 26479.7397964 (1447)	total: 41.2s	remaining: 8.32s
1529:	learn: 9158.8087356	test: 26509.2847502	best: 26479.7397964 (1447)	total: 41.2s	remaining: 8.29s
1530:	learn: 9157.3691940	test: 26510.8102101	best: 26479.7397964 (1447)	total: 41.2s	remaining: 8.26s
1531:	learn: 9154.7893242	test: 26510.7007490	best: 26479.7397964 (1447)	total: 41.2s	remaining: 8.24s
1532:	learn: 9152.6743438	test: 26510.0072569	best: 26479.7397964 (1447)	tot

1609:	learn: 8927.0120930	test: 26489.0911237	best: 26479.7397964 (1447)	total: 43.5s	remaining: 6.15s
1610:	learn: 8926.3389057	test: 26486.9281600	best: 26479.7397964 (1447)	total: 43.5s	remaining: 6.13s
1611:	learn: 8916.6474023	test: 26486.0034277	best: 26479.7397964 (1447)	total: 43.6s	remaining: 6.11s
1612:	learn: 8910.9960821	test: 26483.6069991	best: 26479.7397964 (1447)	total: 43.6s	remaining: 6.08s
1613:	learn: 8907.9893487	test: 26486.0103647	best: 26479.7397964 (1447)	total: 43.6s	remaining: 6.05s
1614:	learn: 8906.4385641	test: 26485.5252864	best: 26479.7397964 (1447)	total: 43.7s	remaining: 6.03s
1615:	learn: 8902.2354733	test: 26490.5864875	best: 26479.7397964 (1447)	total: 43.7s	remaining: 6s
1616:	learn: 8902.1573124	test: 26490.5486234	best: 26479.7397964 (1447)	total: 43.7s	remaining: 5.98s
1617:	learn: 8902.0409234	test: 26490.5470469	best: 26479.7397964 (1447)	total: 43.8s	remaining: 5.95s
1618:	learn: 8899.4308514	test: 26487.7039767	best: 26479.7397964 (1447)	tot

1689:	learn: 8672.5864567	test: 26473.7900244	best: 26470.3367466 (1645)	total: 46s	remaining: 4.03s
1690:	learn: 8667.3639935	test: 26478.5916574	best: 26470.3367466 (1645)	total: 46.1s	remaining: 4s
1691:	learn: 8662.5414475	test: 26475.1795391	best: 26470.3367466 (1645)	total: 46.1s	remaining: 3.98s
1692:	learn: 8660.0336549	test: 26477.2510374	best: 26470.3367466 (1645)	total: 46.1s	remaining: 3.95s
1693:	learn: 8655.1631721	test: 26479.4208776	best: 26470.3367466 (1645)	total: 46.1s	remaining: 3.92s
1694:	learn: 8645.7504925	test: 26479.9610299	best: 26470.3367466 (1645)	total: 46.2s	remaining: 3.89s
1695:	learn: 8645.2120697	test: 26480.6292889	best: 26470.3367466 (1645)	total: 46.2s	remaining: 3.87s
1696:	learn: 8644.1399204	test: 26480.5196199	best: 26470.3367466 (1645)	total: 46.2s	remaining: 3.84s
1697:	learn: 8643.7955789	test: 26479.9687575	best: 26470.3367466 (1645)	total: 46.2s	remaining: 3.81s
1698:	learn: 8643.2546356	test: 26480.6180646	best: 26470.3367466 (1645)	total

1771:	learn: 8449.7955782	test: 26448.6420068	best: 26446.1365238 (1733)	total: 48.4s	remaining: 1.8s
1772:	learn: 8444.1180788	test: 26445.9397777	best: 26445.9397777 (1772)	total: 48.5s	remaining: 1.78s
1773:	learn: 8439.8500002	test: 26450.7833866	best: 26445.9397777 (1772)	total: 48.5s	remaining: 1.75s
1774:	learn: 8438.2445576	test: 26448.8148027	best: 26445.9397777 (1772)	total: 48.5s	remaining: 1.72s
1775:	learn: 8436.5062433	test: 26449.7080352	best: 26445.9397777 (1772)	total: 48.5s	remaining: 1.69s
1776:	learn: 8428.7455449	test: 26459.3224729	best: 26445.9397777 (1772)	total: 48.6s	remaining: 1.67s
1777:	learn: 8425.3406567	test: 26461.8541363	best: 26445.9397777 (1772)	total: 48.6s	remaining: 1.64s
1778:	learn: 8422.5078731	test: 26461.1964389	best: 26445.9397777 (1772)	total: 48.6s	remaining: 1.61s
1779:	learn: 8418.2878204	test: 26460.1381231	best: 26445.9397777 (1772)	total: 48.6s	remaining: 1.58s
1780:	learn: 8416.5670306	test: 26463.4057569	best: 26445.9397777 (1772)	t

## Drop the same columns as before except this time no SalesPrice.

In [28]:
test_2 = test.drop(["Id","Alley","PoolQC","Fence","MiscFeature"], axis=1)

## Here we predict the results using the dataframe of the test.csv dataset. After that you create the prediction results using the housing sales_price_solution.csv file.
## A Dataframe is just what you get when you read in a dataset like train.csv into the Jupyter notebook like we did in the second code-cell in this notebook.

In [30]:
result = cat_model.predict(test_2)
sub['SalesPrice'] = result
sub.to_csv('sales_price_solution.csv', index = 0)